pkg/webhook/admission.go

```diff
@@ -0,0 +1,230 @@
+	admissionregistrationv1beta1 "k8s.io/api/admissionregistration/v1beta1"
+	v1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+
+	"k8s.io/client-go/kubernetes"
+)
+
+var (
+	runtimeScheme = runtime.NewScheme()
+	codecs        = serializer.NewCodecFactory(runtimeScheme)
+	deserializer  = codecs.UniversalDeserializer()
+
+	// (https://github.com/kubernetes/kubernetes/issues/57982)
+	defaulter = runtime.ObjectDefaulter(runtimeScheme)
+)
+
+var (
+	ignoredNamespaces = []string{
+		metav1.NamespaceSystem,
+		metav1.NamespacePublic,
+	}
+)
+
+const (
+	admissionWebhookAnnotationValidateKey = "admission-webhook-openebs.io/validate"
+	admissionWebhookAnnotationMutateKey   = "admission-webhook-openebs.io/mutate"
+	admissionWebhookAnnotationStatusKey   = "admission-webhook-openebs.io/status"
+)
+
+type WebhookServer struct {
+	Server *http.Server
+
+	// kubeClient is a standard kubernetes clientset
+	KubeClient kubernetes.Interface
+
+	// Clientset is a openebs custom resource package generated for custom API group.
+	Clientset clientset.Interface
+}
+
+// WhSvrParameters are webhook parameters are server configures parameters
+type WhSvrParameters struct {
+	Port     int    // webhook server port
+	CertFile string // path to the x509 certificate for https
+	KeyFile  string // path to the x509 private key matching `CertFile`
+}
+
+func init() {
+	_ = corev1.AddToScheme(runtimeScheme)
+	_ = admissionregistrationv1beta1.AddToScheme(runtimeScheme)
+	// defaulting with webhooks:
+	// https://github.com/kubernetes/kubernetes/issues/57982
+	_ = v1.AddToScheme(runtimeScheme)
+}
+
+func admissionRequired(ignoredList []string, admissionAnnotationKey string, metadata *metav1.ObjectMeta) bool {
+	// skip special kubernetes system namespaces
+	for _, namespace := range ignoredList {
+		if metadata.Namespace == namespace {
+			glog.Infof("Skip validation for %v for it's in special namespace:%v", metadata.Name, metadata.Namespace)
+			return false
+		}
+	}
+
+	annotations := metadata.GetAnnotations()
+	if annotations == nil {
+		annotations = map[string]string{}
+	}
+
+	var required bool
+	switch strings.ToLower(annotations[admissionAnnotationKey]) {
+	default:
+		required = true
+	case "n", "no", "false", "off":
+		required = false
+	}
+	return required
+}
+
+func validationRequired(ignoredList []string, metadata *metav1.ObjectMeta) bool {
+	required := admissionRequired(ignoredList, admissionWebhookAnnotationValidateKey, metadata)
+	glog.Infof("Validation policy for %v/%v: required:%v", metadata.Namespace, metadata.Name, required)
+	return required
+}
+
+// validate deployments and services
+func (whsvr *WebhookServer) validate(ar *v1beta1.AdmissionReview) *v1beta1.AdmissionResponse {
+	req := ar.Request
+	status := &v1beta1.AdmissionResponse{}
+
+	if ar.Request.Operation != v1beta1.Delete && req.Kind.Kind != "PersistentVolumeClaim" {
```

Comment:
this is supposed to be `!Delete || !PVC`

Reply:
or make it as !(Delete && PVC) for code readability.. also, please add comments

Reply:
@vishnuitta the changes has been already updated ... these were  the older commits  :slightly_smiling_face: .